relayer/processor/path_processor_internal.go

```diff
@@ -766,38 +795,55 @@ func (pp *PathProcessor) assembleAndSendMessages(
 	// Each assembleMessage call below will make a query on the source chain, so these operations can run in parallel.
 	var wg sync.WaitGroup
 
-	for i, msg := range messages.packetMessages {
-		wg.Add(1)
-		go pp.assembleMessage(ctx, msg, src, dst, &om, i, &wg)
-	}
-
+	// connection messages are highest priority
+	om.connMsgs = make([]connectionMessageToTrack, len(messages.connectionMessages))
 	for i, msg := range messages.connectionMessages {
 		wg.Add(1)
 		go pp.assembleMessage(ctx, msg, src, dst, &om, i, &wg)
 	}
 
-	for i, msg := range messages.channelMessages {
-		wg.Add(1)
-		go pp.assembleMessage(ctx, msg, src, dst, &om, i, &wg)
+	wg.Wait()
+
+	if len(om.msgs) == 1 {
+		om.chanMsgs = make([]channelMessageToTrack, len(messages.channelMessages))
+		// only assemble and send channel handshake messages if there are no conn handshake messages
+		// this prioritizes connection handshake messages, useful if a connection handshake needs to occur before a channel handshake
+		for i, msg := range messages.channelMessages {
+			wg.Add(1)
+			go pp.assembleMessage(ctx, msg, src, dst, &om, i, &wg)
+		}
+
+		wg.Wait()
 	}
 
-	wg.Wait()
+	if len(om.msgs) == 1 {
+		om.pktMsgs = make([]packetMessageToTrack, len(messages.packetMessages))
+		// only assemble and send packet messages if there are no handshake or ICQ messages
```

Comment:
Wouldn't the ICQ messages fall into packet messages also? It's not clear to me what the distinction is here between say a token transfer packet and an icq packet.

Reply:
This comment is stale. ICQ handling only exists in https://github.com/cosmos/relayer/pull/954